cmd/cluster-replication.go

```
@@ -0,0 +1,1278 @@
+
+			// Use a helper to generate unique priority numbers.
+			Priority: fmt.Sprintf("%d", getPriorityHelper(replicationConfig)),
+
+			Op:         replication.AddOption,
+			RuleStatus: "enable",
+			DestBucket: targetARN,
+
+			// Replicate everything!
+			ReplicateDeletes:        "enable",
+			ReplicateDeleteMarkers:  "enable",
+			ReplicaSync:             "enable",
+			ExistingObjectReplicate: "enable",
+		})
+		if err != nil {
+			fmt.Printf("%s->%s: rule add err: %#v\n", c.state.Name, peer.Name, err)
+			return err
+		}
+		// Now convert the configuration back to server's type so we can
+		// do some validation.
+		newReplCfgBytes, err := xml.Marshal(replicationConfig)
+		if err != nil {
+			return wrapCRErr(err)
+		}
+		newReplicationConfig, err := sreplication.ParseConfig(bytes.NewReader(newReplCfgBytes))
+		if err != nil {
+			return wrapCRErr(err)
+		}
+		sameTarget, apiErr := validateReplicationDestination(ctx, bucket, newReplicationConfig)
+		if apiErr != noError {
+			return wrapCRErr(fmt.Errorf("bucket replication config validation error: %#v", apiErr))
+		}
+		err = newReplicationConfig.Validate(bucket, sameTarget)
+		if err != nil {
+			return wrapCRErr(err)
+		}
+		// Config looks good, so we save it.
+		replCfgData, err := xml.Marshal(newReplicationConfig)
+		if err != nil {
+			return wrapCRErr(err)
+		}
+		err = globalBucketMetadataSys.Update(bucket, bucketReplicationConfig, replCfgData)
+		if err != nil {
+			fmt.Printf("%s->%s: replconferr: %v\n", c.state.Name, peer.Name, err)
+			return wrapCRErr(err)
+		}
+	}
+	return nil
+}
+
+// PeerBucketDeleteHandler - deletes bucket on local in response to a delete
+// bucket request from a peer.
+func (c *ClusterReplMgr) PeerBucketDeleteHandler(ctx context.Context, bucket string, forceDelete bool) error {
+	c.RLock()
+	defer c.RUnlock()
+	if !c.Enabled {
+		return errCRNotEnabled
+	}
+
+	// FIXME: need to handle cases where globalDNSConfig is set.
+
+	objAPI := newObjectLayerFn()
```

Let's return an error if objAPI is nil

---

This is done now.